Add tests

domcorvasce opened the pull request.

Commit:
Add Mocha and Chai for unit testing

Commit:
Add test case for Card

Commit:
Add test case for Player

Commit:
Update package.json

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+const expect = require('./../helper').expect;
```
what's the difference in doing `require('chai').expect`?

domcorvasce replied:
Fixed

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+const expect = require('./../helper').expect;
+
+const Card = require('./../../src/core/card');
```
Usually you don't put the .spec.js files in a different folder to avoid this (`../../`): they should go in the same folder of the modules. Btw `./../` is the same as `../`

domcorvasce replied:
Fixed

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+
+describe('Card', () => {
+  beforeEach(() => {
+    this.card = new Card(2, 'spades');
```
`this` here is misleading: use a normal variable

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+    this.card = new Card(2, 'spades');
+  });
+
+  describe('#constructor', () => {
```
The `describe` block should contain semantic concepts, not functions.

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+  describe('#constructor', () => {
+    it('initializes a new card', () => {
+      expect(this.card.value).to.equal(2);
+      expect(this.card.seed).to.equal('spades');
```
For what I see here the test consists in checking that the card has always value = 2 and seed = spades. Instantiate the class in the `it` block to make it more clear. 

Also you shouldn't use always use the same instance, you should continuously change the value of the constructor to make the tests more generic. See [triangulation](http://elnur.pro/triangulation-in-testing)

mariolamacchia replied:
Also, when you're using constants in testings (like `2` or `spades`), make sure that you assign it to a variable that you always refer to: if I want to change `2` to `3` I should do it only in one place

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,22 @@
+    it('returns a string representation of card', () => {
+      expect(this.card.toString()).to.equal('2spades');
+    });
+  });
```
See above

mariolamacchia commented on test/core/player.spec.js:
```diff
@@ -0,0 +1,16 @@
+      expect(this.player.points).to.equal(0);
+    });
+  });
+});
```
See above

mariolamacchia commented on package.json:
```diff
@@ -4,11 +4,15 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "./node_modules/mocha/bin/_mocha test/*"
```
Not `./node_modules/mocha/bin/_mocha`, just `mocha`

domcorvasce replied:
I'll fix it ASAP

Commit:
Refactor(tests): Move tests' variables outside beforeEach

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -3,20 +3,15 @@ const expect = require('./../helper').expect;
-    this.card = new Card(2, 'spades');
-  });
+  const cardData = [2, 'spades'];
+  const card = new Card(...cardData);
```
What changes from the previous version?

domcorvasce replied:
Now you can edit card data in one place

mariolamacchia replied:
You're still instantiating the card once for all the tests. All the code must be always in a block (`it`, `beforeEach`, ...): now each test isn't isolated because you use the same instance in each test. 

You have to instantiate a different card in each `it` block

domcorvasce replied:
Oh ok.

domcorvasce replied:
I fixed it moving variables' definition to `beforeEach` so they're defined before each test.

mariolamacchia commented on test/core/card.spec.js:
```diff
@@ -0,0 +1,17 @@
+  });
+
+  it('returns a string representation of card', () => {
+    expect(card.toString()).to.equal(cardData.join(''));
```
When testing it's better to use explicit values, since it has to be very readable

domcorvasce replied:
I  don't understand what you mean.
Should I write '2spades' instead  of `card.toString`?

mariolamacchia replied:
instead of `cardData.join('')`

mariolamacchia commented on test/core/player.spec.js:
```diff
@@ -0,0 +1,12 @@
+const Player = require('./../../src/core/player');
+
+describe('Player', () => {
+  const player = new Player('Dom');
```
same as card, move it to the `it` block

domcorvasce replied:
Fixed

domcorvasce replied:
Fixed

Commit:
Refactor(tests): Move test's variable into 'it' block

Commit:
Refactor(tests): Remove helper.js

Commit:
Refactor(tests): Move tests in the same folder of modules

Commit:
Refactor(tests): Move test's variables declaration to beforeEach

Commit:
Refactor(tests): Remove test folder

Commit:
Refactor(tests): Move test's variables inside it blocks

Commit:
Refactor(tests): Move test's variables to it blocks in player.spec.js too

Commit:
Add a test case to card.spec.js

The added test case check that the Card's constructor
take only the first two arguments passed to it, no more.

mariolamacchia commented on package.json:
```diff
@@ -4,11 +4,15 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "mocha $(find src -name '*.spec.js')"
```
what if I have Windows?

domcorvasce replied:
Change OS

mariolamacchia replied:
`mocha "src/**/*.spec.js`

mariolamacchia commented on src/core/card.spec.js:
```diff
@@ -0,0 +1,27 @@
+describe('Card', () => {
+  it('initializes a new card', () => {
+    const cardData = [2, 'spades'];
+    const card = new Card(...cardData);
```
Do `new Card(2, 'spades')`

domcorvasce replied:
Ok

Commit:
Refactor(tests): Remove cardData

Commit:
Change script for test
